File: src/lib/types/vibration.ts
/**
 * 振動資料核心型別定義
 *
 * 定義 CSV 欄位映射、預覽、資料集及時序片段等 IPC 傳輸的資料結構。
 * 所有 interface 與 Rust 後端 models/vibration.rs 的 serde 序列化結果一一對應。
 */

/** CSV 欄位映射 — 使用者在 ColumnMappingDialog 中選擇時間欄和資料欄後產生 */
export interface ColumnMapping {
	/** 時間軸欄位名稱（epoch seconds 或可解析的 datetime 字串） */
	time_column: string;
	/** 使用者選取的數值資料欄位名稱列表（動態，非固定 x/y/z） */
	data_columns: string[];
}

/** CSV 檔案預覽 — preview_csv_columns 回傳，用於 ColumnMappingDialog 顯示 */
export interface CsvPreview {
	/** CSV 檔案的完整路徑 */
	file_path: string;
	/** CSV 所有欄位名稱（保持檔案中的原始順序） */
	columns: string[];
	/** 資料列總數（不含 header） */
	row_count: number;
}

/** 已載入的振動資料集元資料 — load_vibration_data 回傳 */
export interface VibrationDataset {
	/** 後端產生的唯一識別碼（UUID） */
	id: string;
	/** CSV 檔案的完整路徑 */
	file_path: string;
	/** 顯示用檔名（不含路徑） */
	file_name: string;
	/** 資料集內的總資料點數 */
	total_points: number;
	/** 時間範圍 [起始秒, 結束秒]（epoch seconds） */
	time_range: [number, number];
	/** 載入時使用的欄位映射設定 */
	column_mapping: ColumnMapping;
}

/** 時序資料片段 — get_timeseries_chunk 回傳，供 ECharts 繪製 */
export interface TimeseriesChunk {
	/** 時間戳陣列（epoch seconds），與 channels 中每個陣列等長且索引對齊 */
	time: number[];
	/** 各通道的數值陣列，key 為通道名稱（保持 IndexMap 順序） */
	channels: Record<string, number[]>;
	/** 是否經 LTTB 降採樣（true 表示非完整資料） */
	is_downsampled: boolean;
	/** 降採樣前的原始資料點數（用於 UI 顯示 "N / M points"） */
	original_count: number;
}

File: src-tauri/src/models/vibration.rs
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColumnMapping {
    pub time_column: String,
    pub data_columns: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CsvPreview {
    pub file_path: String,
    pub columns: Vec<String>,
    pub row_count: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VibrationDataset {
    pub id: String,
    pub file_path: String,
    pub file_name: String,
    pub total_points: usize,
    pub time_range: (f64, f64),
    pub column_mapping: ColumnMapping,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimeseriesChunk {
    pub time: Vec<f64>,
    pub channels: IndexMap<String, Vec<f64>>,
    pub is_downsampled: bool,
    pub original_count: usize,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_column_mapping_serialization() {
        let mapping = ColumnMapping {
            time_column: "datetime".into(),
            data_columns: vec!["accel_x".into(), "accel_y".into(), "accel_z".into()],
        };
        let json = serde_json::to_string(&mapping).unwrap();
        let deserialized: ColumnMapping = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.time_column, "datetime");
        assert_eq!(deserialized.data_columns.len(), 3);
        assert_eq!(deserialized.data_columns[2], "accel_z");
    }

    #[test]
    fn test_csv_preview_serialization() {
        let preview = CsvPreview {
            file_path: "/data/sensor.csv".into(),
            columns: vec!["time".into(), "x".into(), "y".into()],
            row_count: 50000,
        };
        let json = serde_json::to_string(&preview).unwrap();
        let deserialized: CsvPreview = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.file_path, "/data/sensor.csv");
        assert_eq!(deserialized.columns.len(), 3);
        assert_eq!(deserialized.row_count, 50000);
    }

    #[test]
    fn test_timeseries_chunk_serialization_with_indexmap_channels() {
        let mut channels = IndexMap::new();
        channels.insert("accel_x".to_string(), vec![0.1, 0.2, 0.3]);
        channels.insert("accel_y".to_string(), vec![1.1, 1.2, 1.3]);

        let chunk = TimeseriesChunk {
            time: vec![1000.0, 1001.0, 1002.0],
            channels,
            is_downsampled: false,
            original_count: 3,
        };

        let json = serde_json::to_string(&chunk).unwrap();
        let deserialized: TimeseriesChunk = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.time.len(), 3);
        assert_eq!(deserialized.channels.len(), 2);
        assert_eq!(deserialized.channels["accel_x"], vec![0.1, 0.2, 0.3]);
        assert_eq!(deserialized.channels["accel_y"], vec![1.1, 1.2, 1.3]);
        assert!(!deserialized.is_downsampled);
        assert_eq!(deserialized.original_count, 3);
    }

    #[test]
    fn test_timeseries_chunk_channel_order_preserved() {
        // IndexMap preserves insertion order
        let mut channels = IndexMap::new();
        channels.insert("z".to_string(), vec![1.0]);
        channels.insert("a".to_string(), vec![2.0]);
        channels.insert("m".to_string(), vec![3.0]);

        let chunk = TimeseriesChunk {
            time: vec![0.0],
            channels,
            is_downsampled: true,
            original_count: 100,
        };

        let json = serde_json::to_string(&chunk).unwrap();
        let deserialized: TimeseriesChunk = serde_json::from_str(&json).unwrap();

        let keys: Vec<&String> = deserialized.channels.keys().collect();
        assert_eq!(keys, vec!["z", "a", "m"]);
    }

    #[test]
    fn test_vibration_dataset_serialization() {
        let dataset = VibrationDataset {
            id: "ds-001".into(),
            file_path: "/data/test.csv".into(),
            file_name: "test.csv".into(),
            total_points: 100000,
            time_range: (1700000000.0, 1700003600.0),
            column_mapping: ColumnMapping {
                time_column: "time".into(),
                data_columns: vec!["x".into()],
            },
        };
        let json = serde_json::to_string(&dataset).unwrap();
        let deserialized: VibrationDataset = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.id, "ds-001");
        assert_eq!(deserialized.total_points, 100000);
        assert!((deserialized.time_range.0 - 1700000000.0).abs() < 1e-6);
    }

    #[test]
    fn test_column_mapping_empty_data_columns() {
        let mapping = ColumnMapping {
            time_column: "ts".into(),
            data_columns: vec![],
        };
        let json = serde_json::to_string(&mapping).unwrap();
        let deserialized: ColumnMapping = serde_json::from_str(&json).unwrap();
        assert!(deserialized.data_columns.is_empty());
    }
}
